src/Container/ApplicationConfigInjectionDelegator.php

```diff
@@ -0,0 +1,373 @@
+            return $application;
+        }
+
+        $config = $container->get('config');
+        if (! isset($config['routes']) && ! isset($config['middleware_pipeline'])) {
+            return $application;
+        }
+
+        self::injectPipelineFromConfig($application, $config);
+        self::injectRoutesFromConfig($application, $config);
+
+        return $application;
+    }
+
+    /**
+     * Inject a middleware pipeline from the middleware_pipeline configuration.
+     *
+     * Inspects the configuration provided to determine if a middleware pipeline
+     * exists to inject in the application.
+     *
+     * If no pipeline is defined, but routes *are*, then the method will inject
+     * the routing and dispatch middleware.
+     *
+     * Use the following configuration format:
+     *
+     * <code>
+     * return [
+     *     'middleware_pipeline' => [
+     *         // An array of middleware to register with the pipeline.
+     *         // entries to register prior to routing/dispatching...
+     *         // - entry for \Zend\Expressive\Router\Middleware\RouteMiddleware::class
+     *         // - entry for \Zend\Expressive\Router\Middleware\DispatchMiddleware::class
+     *         // entries to register after routing/dispatching...
+     *     ],
+     * ];
+     * </code>
+     *
+     * Each item in the middleware_pipeline array must be of the following
+     * specification:
+     *
+     * <code>
+     * [
+     *     // required:
+     *     'middleware' => 'Name of middleware service, or a callable',
+     *     // optional:
+     *     'path'  => '/path/to/match',
+     *     'priority' => 1, // integer
+     * ]
+     * </code>
+     *
+     * Note that the `path` element can only be a literal.
+     *
+     * `priority` is used to shape the order in which middleware is piped to the
+     * application. Values are integers, with high values having higher priority
+     * (piped earlier), and low/negative values having lower priority (piped last).
+     * Default priority if none is specified is 1. Middleware with the same
+     * priority are piped in the order in which they appear.
+     *
+     * Middleware piped may be either callables or service names.
+     *
+     * Additionally, you can specify an array of callables or service names as
+     * the `middleware` value of a specification. Internally, this will create
+     * a `Zend\Stratigility\MiddlewarePipe` instance, with the middleware
+     * specified piped in the order provided.
+     *
+     * @return void
+     */
+    public static function injectPipelineFromConfig(Application $application, array $config)
+    {
+        if (empty($config['middleware_pipeline'])) {
+            if (! isset($config['routes']) || ! is_array($config['routes'])) {
+                return;
+            }
+
+            self::pipeRoutingMiddleware($application);
+            self::pipeDispatchMiddleware($application);
+            return;
+        }
```

Comment:
I'm thinking the above clause should likely be removed. It creates a potentially dangerous situation if the user is creating a programmatic pipeline, but using this delegator, as the above will pipe the routing and dispatch middleware _before_ the user has had a chance to setup the pipeline.

I'm thinking the above should only trigger if the "programmatic pipeline" flag is _not_ enabled.